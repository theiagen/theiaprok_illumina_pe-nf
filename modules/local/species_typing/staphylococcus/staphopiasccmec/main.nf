process STAPHOPIASCCMEC {
    tag "$meta.id"
    label "process_low"

    container "us-docker.pkg.dev/general-theiagen/biocontainers/staphopia-sccmec:1.0.0--hdfd78af_0"

    input:
    tuple val(meta), path(assembly)

    output:
    tuple val(meta), path("*.staphopia-sccmec.summary.tsv") , emit: staphopia_results_tsv
    tuple val(meta), path("*.staphopia-sccmec.hamming.tsv") , emit: staphopia_hamming_distance_tsv
    tuple val(meta), path("TYPES_AND_MECA.txt")             , emit: staphopia_types_and_meca_presence
    path "versions.yml"                                     , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    
    """
    # Run staphopia-sccmec on input assembly (hamming option OFF)
    # Outputs are true/false for each SCCmec type
    staphopia-sccmec \\
        --assembly ${assembly} \\
        ${args} \\
        > ${prefix}.staphopia-sccmec.summary.tsv
    
    # Run staphopia-sccmec on input assembly (hamming option ON)  
    # Outputs are the hamming distance; 0 is exact match
    staphopia-sccmec \\
        --hamming \\
        --assembly ${assembly} \\
        ${args} \\
        > ${prefix}.staphopia-sccmec.hamming.tsv
    
    # please excuse this ugly bash code below :)

    # parse output summary TSV for true matches
    # look for columns that contain the word "True" and print the column numbers in a list to a file col_headers.txt
    awk '{ for (i=1; i<=NF; ++i) { if (\$i ~ "True") print i } }' ${prefix}.staphopia-sccmec.summary.tsv | tee col_headers.txt

    # use column number list to print column headers (example: IV, mecA, etc.) to a file type.txt
    cat col_headers.txt | while read -r COL_NUMBER; do \
      cut -f "\$COL_NUMBER" ${prefix}.staphopia-sccmec.summary.tsv | head -n 1 >>type.txt
      echo "," >>type.txt
    done

    # remove newlines, remove trailing comma; generate output string of comma separated values
    cat type.txt | tr -d '\n' | sed 's|.\$||g' | tee TYPES_AND_MECA.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        staphopia-sccmec: \$(staphopia-sccmec --version 2>&1 | sed 's/^.*staphopia-sccmec //')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.staphopia-sccmec.summary.tsv
    touch ${prefix}.staphopia-sccmec.hamming.tsv
    touch TYPES_AND_MECA.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        staphopia-sccmec: \$(staphopia-sccmec --version 2>&1 | sed 's/^.*staphopia-sccmec //')
    END_VERSIONS
    """
}