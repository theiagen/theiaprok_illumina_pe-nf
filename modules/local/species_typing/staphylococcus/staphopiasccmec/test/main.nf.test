nextflow_process {
    name "Test Process STAPHOPIASCCMEC"
    script "../main.nf"
    process "STAPHOPIASCCMEC"

    test("STAPHOPIASCCMEC - test") {
        setup {
            run("GUNZIP") {
                script "../../../../../nf-core/gunzip/main.nf"
                process {
                    """
                    input[0] = [
                        [id:'test_sample'],
                        file('https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/001/839/195/GCF_001839195.1_ASM183919v1/GCF_001839195.1_ASM183919v1_genomic.fna.gz', checkIfExists: true)
                    ]
                    """
                }
            }
        }
        when {
            process {
                """
                input[0] = GUNZIP.out.gunzip
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.staphopia_results_tsv,
                    process.out.staphopia_hamming_distance_tsv,
                    process.out.staphopia_types_and_meca_presence,
                    process.out.versions
                ).match() }
            )
        }
    }

    test("STAPHOPIASCCMEC  - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    file('https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/001/839/195/GCF_001839195.1_ASM183919v1/GCF_001839195.1_ASM183919v1_genomic.fna.gz', checkIfExists: true)
                ])
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}