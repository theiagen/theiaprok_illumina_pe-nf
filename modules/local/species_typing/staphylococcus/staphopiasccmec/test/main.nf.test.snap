{
    "STAPHOPIASCCMEC  - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.staphopia-sccmec.summary.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.staphopia-sccmec.hamming.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "TYPES_AND_MECA.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    "versions.yml:md5,efa986cf837fac5b6e3bdd3dab8b424c"
                ],
                "staphopia_hamming_distance_tsv": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.staphopia-sccmec.hamming.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "staphopia_results_tsv": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.staphopia-sccmec.summary.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "staphopia_types_and_meca_presence": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "TYPES_AND_MECA.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,efa986cf837fac5b6e3bdd3dab8b424c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.5"
        },
        "timestamp": "2025-06-25T08:50:27.085125"
    },
    "STAPHOPIASCCMEC - test": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.staphopia-sccmec.summary.tsv:md5,fa2f9b997dcb27e4969fe695abc7ad32"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.staphopia-sccmec.hamming.tsv:md5,75489ab577d33e311d78ede0d40d0d2e"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "TYPES_AND_MECA.txt:md5,646c4b9f161e16c64773fcae1485d00c"
                ]
            ],
            [
                "versions.yml:md5,e548a4ec886b19e5055de4f09e7ba8b2"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.5"
        },
        "timestamp": "2025-06-25T08:50:22.766939"
    }
}